models/image.py

```diff
@@ -1242,6 +1077,244 @@ def load(self):
+
+        with SmartSession(session) as session:
+            im_ids = [im.id for im in self.upstream_images]
+
+            # get all the products for all images at once
+            sources_result = session.scalars(
+                sa.select(SourceList).where(
+                    SourceList.image_id.in_(im_ids),
+                    SourceList.provenance_id.in_(prov_ids),
+                )
+            ).all()
+            sources_ids = [s.id for s in sources_result]
+
+            psf_results = session.scalars(
+                sa.select(PSF).where(
+                    PSF.image_id.in_(im_ids),
+                    PSF.provenance_id.in_(prov_ids),
+                )
+            ).all()
+
+            wcs_results = session.scalars(
+                sa.select(WorldCoordinates).where(
+                    WorldCoordinates.source_list_id.in_(sources_ids),
+                    WorldCoordinates.provenance_id.in_(prov_ids),
+                )
+            ).all()
+
+            zp_results = session.scalars(
+                sa.select(ZeroPoint).where(
+                    ZeroPoint.source_list_id.in_(sources_ids),
+                    ZeroPoint.provenance_id.in_(prov_ids),
+                )
+            ).all()
+
+            for im in self.upstream_images:
+                sources = [s for s in sources_result if s.image_id == im.id]  # only get the sources for this image
+                if len(sources) > 1:
+                    raise ValueError(
+                        f"Image {im.id} has more than one SourceList matching upstream provenance."
+                    )
+                elif len(sources) == 1:
+                    im.sources = sources[0]
+
+                psfs = [p for p in psf_results if p.image_id == im.id]  # only get the psfs for this image
+                if len(psfs) > 1:
+                    raise ValueError(
+                        f"Image {im.id} has more than one PSF matching upstream provenance."
+                    )
+                elif len(psfs) == 1:
+                    im.psf = psfs[0]
+
+                wcses = [w for w in wcs_results if w.image_id == im.id]  # only get the wcses for this image
+                if len(wcses) > 1:
+                    raise ValueError(
+                        f"Image {im.id} has more than one WCS matching upstream provenance."
+                    )
+                elif len(wcses) == 1:
+                    im.wcs = wcses[0]
+
+                zps = [z for z in zp_results if z.image_id == im.id]  # only get the zps for this image
+                if len(zps) > 1:
+                    raise ValueError(
+                        f"Image {im.id} has more than one ZeroPoint matching upstream provenance."
+                    )
+                elif len(zps) == 1:
+                    im.zp = zps[0]
+
+    def get_upstreams(self, session=None):
+        """
+        Get the upstream images and associated products that were used to make this image.
```

Comment:
This description isn't really right.  The sources, psf, wcs, etc. will only be loaded for the upstream images if the upstream images are already in the session and you've called something to fill in the .sources, .zp, etc. fields.